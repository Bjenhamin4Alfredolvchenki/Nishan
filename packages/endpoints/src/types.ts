import {
	TBlock,
	ICollection,
	TView,
	ISpace,
	INotionUser,
	ISpaceView,
	IUserRoot,
	IUserSettings,
	TDataType
} from '@nishans/types';

export type UpdateCacheManuallyParam = (string | [string, TDataType])[];

export interface ICache {
	block: Map<string, TBlock>;
	collection: Map<string, ICollection>;
	collection_view: Map<string, TView>;
	space: Map<string, ISpace>;
	notion_user: Map<string, INotionUser>;
	space_view: Map<string, ISpaceView>;
	user_root: Map<string, IUserRoot>;
	user_settings: Map<string, IUserSettings>;
}

export interface CtorArgs {
	token: string;
	interval: number;
	user_id: string;
	shard_id: number;
	space_id: string;
	cache: ICache;
}
