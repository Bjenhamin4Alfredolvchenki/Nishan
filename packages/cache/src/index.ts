export * from './types';
export * from './NotionCacheClass';
export * from './NotionCacheObject';
export * from '../utils';
