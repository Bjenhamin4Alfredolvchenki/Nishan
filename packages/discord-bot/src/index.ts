'use strict';

module.exports = discordBot;

function discordBot() {
    // TODO
}
