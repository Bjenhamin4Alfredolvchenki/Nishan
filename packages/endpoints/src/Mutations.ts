import {
	SetPageNotificationsAsReadParams,
	SetSpaceNotificationsAsReadParams,
	RemoveUsersFromSpaceParams,
	RemoveUsersFromSpaceResult,
	InviteGuestsToSpaceParams,
	CreateSpaceParams,
	CreateSpaceResult,
	SaveTransactionParams,
	EnqueueTaskParams,
	EnqueueTaskResult,
	SetBookmarkMetadataParams,
	InitializePageTemplateResult,
	InitializePageTemplateParams,
	InitializeGoogleDriveBlockParams,
	InitializeGoogleDriveBlockResult
} from '@nishans/types';

import { Configs } from '.';
import { sendRequest } from '../utils';

const Mutations = {
	async setPageNotificationsAsRead (params: SetPageNotificationsAsReadParams, configs?: Partial<Configs>) {
		return await sendRequest('setPageNotificationsAsRead', params, configs);
	},
	async setSpaceNotificationsAsRead (params: SetSpaceNotificationsAsReadParams, configs?: Partial<Configs>) {
		return await sendRequest('setSpaceNotificationsAsRead', params, configs);
	},
	async removeUsersFromSpace (params: RemoveUsersFromSpaceParams, configs?: Partial<Configs>) {
		return await sendRequest<RemoveUsersFromSpaceResult>('removeUsersFromSpace', params, configs);
	},
	async inviteGuestsToSpace (params: InviteGuestsToSpaceParams, configs?: Partial<Configs>) {
		return await sendRequest('inviteGuestsToSpace', params, configs);
	},
	async createSpace (params: CreateSpaceParams, configs?: Partial<Configs>) {
		return await sendRequest<CreateSpaceResult>('createSpace', params, configs);
	},
	async saveTransactions (params: SaveTransactionParams, configs?: Partial<Configs>) {
		return await sendRequest('saveTransactions', params, configs);
	},
	async enqueueTask (params: EnqueueTaskParams, configs?: Partial<Configs>) {
		return await sendRequest<EnqueueTaskResult>('enqueueTask', params, configs);
	},
	async setBookmarkMetadata (params: SetBookmarkMetadataParams, configs?: Partial<Configs>) {
		return await sendRequest('setBookmarkMetadata', params, configs);
	},
	async initializePageTemplate (params: InitializePageTemplateParams, configs?: Partial<Configs>) {
		return await sendRequest<InitializePageTemplateResult>('initializePageTemplate', params, configs);
	},
	async initializeGoogleDriveBlock (params: InitializeGoogleDriveBlockParams, configs?: Partial<Configs>) {
		return await sendRequest<InitializeGoogleDriveBlockResult>('initializeGoogleDriveBlock', params, configs);
	}
};

export default Mutations;
