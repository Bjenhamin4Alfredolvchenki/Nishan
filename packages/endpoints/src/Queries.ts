import {
	GetPageVisitsParams,
	GetPageVisitsResult,
	GetUserSharedPagesParams,
	GetUserSharedPagesResult,
	GetUserTasksResult,
	GetPublicPageDataParams,
	GetPublicPageDataResult,
	GetPublicSpaceDataParams,
	GetPublicSpaceDataResult,
	GetSubscriptionDataParams,
	GetSubscriptionDataResult,
	LoadBlockSubtreeParams,
	LoadBlockSubtreeResult,
	GetGenericEmbedBlockDataParams,
	GetGenericEmbedBlockDataResult,
	GetUploadFileUrlParams,
	GetUploadFileUrlResult,
	GetGoogleDriveAccountsResult,
	GetBackLinksForBlockResult,
	FindUserResult,
	SyncRecordValuesParams,
	SyncRecordValuesResult,
	QueryCollectionParams,
	QueryCollectionResult,
	LoadUserContentResult,
	LoadPageChunkParams,
	LoadPageChunkResult,
	GetSpacesResult,
	GetBackLinksForBlockParams,
	FindUserParams,
	GetJoinableSpacesResult,
	IsUserDomainJoinableResult,
	IsEmailEducationResult,
	GetUserNotificationsResult,
	GetUserNotificationsParams,
	GetTasksParams,
	GetTasksResult,
	RecordPageVisitResult,
	RecordPageVisitParams
} from '@nishans/types';

import { Configs } from '.';
import { sendRequest } from '../utils';

const Queries = {
	async getPageVisits (params: GetPageVisitsParams, configs?: Partial<Configs>) {
		return await sendRequest<GetPageVisitsResult>('getPageVisits', params, configs);
	},

	async getUserSharedPages (params: GetUserSharedPagesParams, configs?: Partial<Configs>) {
		return await sendRequest<GetUserSharedPagesResult>('getUserSharedPages', params, configs);
	},

	async getUserTasks (configs?: Partial<Configs>) {
		return await sendRequest<GetUserTasksResult>('getUserTasks', {}, configs);
	},

	async getPublicPageData (params: GetPublicPageDataParams, configs?: Partial<Configs>) {
		return await sendRequest<GetPublicPageDataResult>('getPublicPageData', params, configs);
	},

	async getPublicSpaceData (params: GetPublicSpaceDataParams, configs?: Partial<Configs>) {
		return await sendRequest<GetPublicSpaceDataResult>('getPublicSpaceData', params, configs);
	},

	async getSubscriptionData (params: GetSubscriptionDataParams, configs?: Partial<Configs>) {
		return await sendRequest<GetSubscriptionDataResult>('getSubscriptionData', params, configs);
	},

	async loadBlockSubtree (params: LoadBlockSubtreeParams, configs?: Partial<Configs>) {
		return await sendRequest<LoadBlockSubtreeResult>('loadBlockSubtree', params, configs);
	},

	async getSpaces (configs?: Partial<Configs>) {
		return await sendRequest<GetSpacesResult>('getSpaces', {}, configs);
	},

	async getGenericEmbedBlockData (params: GetGenericEmbedBlockDataParams, configs?: Partial<Configs>) {
		return await sendRequest<GetGenericEmbedBlockDataResult>('getGenericEmbedBlockData', params, configs);
	},

	async getUploadFileUrl (params: GetUploadFileUrlParams, configs?: Partial<Configs>) {
		return await sendRequest<GetUploadFileUrlResult>('getUploadFileUrl', params, configs);
	},

	async getGoogleDriveAccounts (configs?: Partial<Configs>) {
		return await sendRequest<GetGoogleDriveAccountsResult>('getGoogleDriveAccounts', {}, configs);
	},

	async getBacklinksForBlock (params: GetBackLinksForBlockParams, configs?: Partial<Configs>) {
		return await sendRequest<GetBackLinksForBlockResult>('getBacklinksForBlock', params, configs);
	},

	async findUser (params: FindUserParams, configs?: Partial<Configs>) {
		return await sendRequest<FindUserResult>('findUser', params, configs);
	},

	async syncRecordValues (params: SyncRecordValuesParams, configs?: Partial<Configs>) {
		return await sendRequest<SyncRecordValuesResult>('syncRecordValues', params, configs);
	},

	async queryCollection (params: QueryCollectionParams, configs?: Partial<Configs>) {
		return await sendRequest<QueryCollectionResult>('queryCollection', params, configs);
	},

	async loadUserContent (configs?: Partial<Configs>) {
		return await sendRequest<LoadUserContentResult>('loadUserContent', {}, configs);
	},

	async loadPageChunk (params: LoadPageChunkParams, configs?: Partial<Configs>) {
		return await sendRequest<LoadPageChunkResult>('loadPageChunk', params, configs);
	},

	async recordPageVisit (params: RecordPageVisitParams, configs?: Partial<Configs>) {
		return await sendRequest<RecordPageVisitResult>('recordPageVisit', params, configs);
	},

	async getJoinableSpaces (configs?: Partial<Configs>) {
		return await sendRequest<GetJoinableSpacesResult>('getJoinableSpaces', {}, configs);
	},

	async isUserDomainJoinable (configs?: Partial<Configs>) {
		return await sendRequest<IsUserDomainJoinableResult>('isUserDomainJoinable', {}, configs);
	},

	async isEmailEducation (configs?: Partial<Configs>) {
		return await sendRequest<IsEmailEducationResult>('isEmailEducation', {}, configs);
	},

	async getUserNotifications (params: GetUserNotificationsParams, configs?: Partial<Configs>) {
		return await sendRequest<GetUserNotificationsResult>('getUserNotifications', params, configs);
	},

	async getTasks (params: GetTasksParams, configs?: Partial<Configs>) {
		return await sendRequest<GetTasksResult>('getTasks', params, configs);
	}
};

export default Queries;
