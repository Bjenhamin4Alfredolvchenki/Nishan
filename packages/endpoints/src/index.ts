'use strict';

module.exports = endpoints;

function endpoints() {
    // TODO
}
