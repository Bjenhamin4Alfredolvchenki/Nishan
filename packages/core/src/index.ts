import {NotionCacheClass} from "@nishans/cache";
import { INotionUser } from "@nishans/types";

import Block from "./Block";
import Collection from "./Collection";
import CollectionViewPage from "./CollectionViewPage";
import CollectionView from "./CollectionView";
import NotionUser from "./NotionUser";
import Page from "./Page";
import UserSettings from "./UserSettings";
import UserRoot from "./UserRoot";
import SpaceView from "./SpaceView";
import Space from "./Space";
import SchemaUnit from "./SchemaUnit";
import CollectionBlock from "./CollectionBlock";
import NotionPermissions from "./Permissions";
export * from "./View";
import { constructLogger, error, iterateAndGetChildren, transformToMultiple } from "../utils";
import {Logger, NishanArg,FilterType, FilterTypes} from "../types";

class Nishan extends NotionCacheClass {
  token: string;
  interval: number;
  init_cache: boolean;
  logger: Logger;

  constructor(arg: Pick<NishanArg, "token"> & { interval?: number, logger?: Logger }) {
    super(arg);
    this.token = arg.token;
    this.interval = arg.interval ?? 500;
    this.init_cache = false;
    this.logger = constructLogger(arg.logger)
  }

  #initializeCache = async () => {
    if (!this.init_cache) {
      await this.initializeNotionCache();
      this.init_cache = true;
    }
  }

  /**
   * Get `INotionUser` and return `NotionUser` by the `args` param
   * @param args An string id, a predicate passed the INotionUser or undefined to indicate everything
   */
  async getNotionUser(arg?: FilterType<INotionUser>) {
    return (await this.getNotionUsers(transformToMultiple(arg), false))[0];
  }

  /**
   * Get `INotionUser[]` and return `NotionUser[]` by the `args` param
   * @param args An array of string ids, a predicate passed the INotionUser or undefined to indicate everything
   */
  async getNotionUsers(args?: FilterTypes<INotionUser>, multiple?: boolean) {
    await this.#initializeCache();

    const common_props = {
      token: this.token,
      cache: this.cache,
      interval: this.interval,
      logger: this.logger,
      stack: [],
    }

    const notion_user_ids: string[] = [];
    for (const [notion_user_id] of this.cache.notion_user) {
      notion_user_ids.push(notion_user_id)
    }

    return await iterateAndGetChildren<INotionUser, INotionUser, NotionUser[]>(args, (id)=>this.cache.notion_user.get(id), {
      ...common_props,
      multiple,
      child_ids: notion_user_ids,
      child_type: 'notion_user',
      container: [],
      parent_id: '',
      parent_type: 'notion_user'
    }, (id, _, container)=>container.push(new NotionUser({ ...common_props, user_id: id, id, space_id: "0", shard_id: 0 })));
  }
}

export default Nishan;
export * from "../types";
export * from "../utils";

export {
  Block,
  Collection,
  CollectionViewPage,
  CollectionView,
  NotionUser,
  Page,
  UserSettings,
  UserRoot,
  SpaceView,
  Space,
  SchemaUnit,
  CollectionBlock,
  NotionPermissions
}