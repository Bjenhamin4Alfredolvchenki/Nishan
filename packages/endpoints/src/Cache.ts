import { ICollection, ISpace, ISpaceView, IUserRoot, RecordMap, SyncRecordValues, TBlock, TDataType } from '@nishans/types';
import { validateCache, constructNotionHeaders } from '../utils';
import { getSpaces, queryCollection, syncRecordValues } from '../src';
import { Configs, CtorArgs, ICache, NotionHeaders, UpdateCacheManuallyParam } from './types';

export default class Cache {
	cache: ICache;
	token: string;
	interval: number;
	headers: NotionHeaders;
	user_id: string;

	constructor ({ cache, token, interval, user_id }: Omit<CtorArgs, 'shard_id' | 'space_id'>) {
		this.cache = (cache && validateCache(cache)) || {
			block: new Map(),
			collection: new Map(),
			space: new Map(),
			collection_view: new Map(),
			notion_user: new Map(),
			space_view: new Map(),
			user_root: new Map(),
			user_settings: new Map()
		};
		this.headers = constructNotionHeaders({token, user_id});
		this.token = token;
		this.interval = interval ?? 500;
		this.user_id = user_id ?? '';
	}

	getConfigs = (): Configs => {
		return {
			token: this.token,
			user_id: this.user_id,
			interval: this.interval
		};
	};

	/**
   * Save the passed recordMap to cache
   * @param recordMap RecordMap map to save to cache
   */
	saveToCache (recordMap: Partial<RecordMap>) {
		(Object.keys(this.cache) as (keyof ICache)[]).forEach((key) => {
			if (recordMap[key])
				Object.entries(recordMap[key] || {}).forEach(([ record_id, record_value ]) => {
					this.cache[key].set(record_id, record_value.value);
				});
		});
	}

	returnNonCachedData (ids: UpdateCacheManuallyParam): UpdateCacheManuallyParam {
		return ids.filter(
			(info) => !Boolean(Array.isArray(info) ? this.cache[info[1]].get(info[0]) : this.cache.block.get(info))
		);
  }
  
  async initializeCache(){
    const data = await getSpaces({token: this.token}); 
    const external_notion_users: Set<string> = new Set();
    Object.values(data).forEach(recordMap => {
      Object.values(recordMap.space).forEach(space => space.value.permissions.forEach(permission =>
        permission.user_id && external_notion_users.add(permission.user_id)
      ))
      this.saveToCache(recordMap)
    });

    const { recordMap } = await syncRecordValues({
      requests: Array.from(external_notion_users.values()).map(external_notion_user => ({ table: "notion_user", id: external_notion_user, version: -1 }))
    }, {token: this.token, interval: 0});
    this.saveToCache(recordMap);
  }

	async updateCacheManually (arg: UpdateCacheManuallyParam | string) {
		const sync_record_values: SyncRecordValues[] = [];
		if (Array.isArray(arg))
			arg.forEach((arg: string | [string, TDataType]) => {
				if (Array.isArray(arg)) sync_record_values.push({ id: arg[0], table: arg[1], version: 0 });
				else if (typeof arg === 'string') sync_record_values.push({ id: arg, table: 'block', version: 0 });
			});
    else if (typeof arg === 'string') sync_record_values.push({ id: arg, table: 'block', version: 0 });
    if (sync_record_values.length){
      const data = await syncRecordValues({ requests: sync_record_values }, {token: this.token, interval: 0});
      this.saveToCache(data.recordMap);
    }
	}

	async updateCacheIfNotPresent (arg: UpdateCacheManuallyParam) {
		const sync_record_values: SyncRecordValues[] = [];
		arg.forEach((arg: string | [string, TDataType]) => {
			if (Array.isArray(arg) && !this.cache[arg[1]].get(arg[0]))
				sync_record_values.push({ id: arg[0], table: arg[1], version: 0 });
			else if (typeof arg === 'string' && !this.cache.block.get(arg))
				sync_record_values.push({ id: arg, table: 'block', version: 0 });
		});
		if (sync_record_values.length) {
      const data = await syncRecordValues({ requests: sync_record_values }, {token: this.token, interval: 0});
      this.saveToCache(data.recordMap);
    }
  }
  
  async initializeCacheForSpecificData(id: string, type: TDataType){
    const container: UpdateCacheManuallyParam = [], data = this.cache[type].get(id);
    if (type === "block") {
      const temp_data = data as TBlock;
      if (temp_data.type === "page")
        container.push(...temp_data.content);
      if (temp_data.type === "collection_view" || temp_data.type === "collection_view_page") {
        temp_data.view_ids.map((view_id) => container.push([view_id, "collection_view"]))
        container.push([temp_data.collection_id, "collection"])
      }
    } else if (type === "space") {
      const temp_data = data as ISpace;
      container.push(...temp_data.pages);
      temp_data.permissions.forEach((permission) => container.push([permission.user_id, "notion_user"]))
    } else if (type === "user_root")
      (data as IUserRoot).space_views.map((space_view => container.push([space_view, "space_view"]))) ?? []
    else if (type === "collection") {
      container.push(...((data as ICollection).template_pages ?? []))
      const {recordMap} = await queryCollection({
        collectionId: id,
        collectionViewId: "",
        query: {},
        loader: {
          type: "table",
          loadContentCover: true
        }
      }, this.getConfigs())
      this.saveToCache(recordMap);
    }
    else if (type === "space_view")
      container.push(...(data as ISpaceView).bookmarked_pages ?? [])

    const non_cached = this.returnNonCachedData(container)
    await this.updateCacheManually(non_cached);

    // If the block is a page, for all the collection block contents, fetch the collection attached with it as well
    if(type === "block"){
      const temp_data = data as TBlock;
      if(temp_data.type === "page"){
        const collection_blocks_ids: UpdateCacheManuallyParam = [];
        for (let index = 0; index < temp_data.content.length; index++) {
          const content_id = temp_data.content[index];
          const content = this.cache.block.get(content_id)
          if(content && (content.type === "collection_view_page" || content.type === "collection_view"))
            collection_blocks_ids.push([content.collection_id, "collection"])
        }
        const non_cached = this.returnNonCachedData(collection_blocks_ids)
        await this.updateCacheManually(non_cached);
      }
    }
  }
}
