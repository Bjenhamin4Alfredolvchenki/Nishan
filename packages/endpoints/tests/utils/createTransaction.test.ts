import { createTransaction } from '../../src';

describe.only('createTransaction', () => {
	it(`Should create correct request`, () => {
		const transactions = createTransaction(123, 'space_id', []);
		const data = JSON.parse(JSON.stringify(transactions));
		expect(data.transactions[0]).toEqual(
			expect.objectContaining({
				id: expect.any(String),
				shardId: 123,
				spaceId: 'space_id',
				operations: expect.any(Array)
			})
		);
	});
});
