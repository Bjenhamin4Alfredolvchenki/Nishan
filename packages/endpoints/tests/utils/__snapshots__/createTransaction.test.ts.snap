// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`createTransaction Should create correct request 1`] = `
Object {
  "id": Any<String>,
  "operations": Any<Array>,
  "shardId": 123,
  "spaceId": "space_id",
}
`;
